### ISSUE TITLE
test-cmd.sh is racy

### ISSUE THREAD

**Issue by brendandburns** [kubernetes/kubernetes#14467]
@smarterclayton @deads2k @kubernetes/kubectl 

```
Successful get nodes 127.0.0.1 {{.spec.unschedulable}}: 
Error from server: node "127.0.0.1" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
!!! Error in ./hack/test-cmd.sh:783
  'kubectl patch "${kube_flags[@]}" nodes "127.0.0.1" -p='{"spec":{"unschedulable":true}}'' exited with status 1
Call stack:
  1: ./hack/test-cmd.sh:783 runTests(...)
  2: ./hack/test-cmd.sh:832 main(...)
```

My guess as to what is happening:
   * we do a GET for the node status
   * separately, the node status is updated, resource version changes
   * we compute the new node state with unschedulable = true using what we got from the GET
   * we try to POST with the old resource version and it fails.

The simplest solution is just to wrap the update call in some retry logic.

Longer term, kubectl should probably support the retry itself.

**Comment by smarterclayton** [comment]
There's an open issue for this - we need a precondition patch for the API
in a long run (apply changes Y if object has X currently).  The cheaper
option is to compare spec and metadata for diffs (what's what edit did)
since most edits are status today.  The blind overwrite on retry is fine
for replace and patch, but not for commands that mutate an initial state
and then submit.

On Sep 23, 2015, at 11:58 PM, Brendan Burns <notifications@github.com>
wrote:

@smarterclayton <https://github.com/smarterclayton> @deads2k
<https://github.com/deads2k> @kubernetes/kubectl
<https://github.com/orgs/kubernetes/teams/kubectl>

Successful get nodes 127.0.0.1 {{.spec.unschedulable}}:
Error from server: node "127.0.0.1" cannot be updated: the object has
been modified; please apply your changes to the latest version and try
again
!!! Error in ./hack/test-cmd.sh:783
  'kubectl patch "${kube_flags[@]}" nodes "127.0.0.1"
-p='{"spec":{"unschedulable":true}}'' exited with status 1
Call stack:
  1: ./hack/test-cmd.sh:783 runTests(...)
  2: ./hack/test-cmd.sh:832 main(...)

My guess as to what is happening:

   - we do a GET for the node status
   - separately, the node status is updated, resource version changes
   - we compute the new node state with unschedulable = true using what we
   got from the GET
   - we try to POST with the old resource version and it fails.

The simplest solution is just to wrap the update call in some retry logic.

Longer term, kubectl should probably support the retry itself.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/14467>.

**Comment by brendandburns** [comment]
@smarterclayton I looked, I couldn't find the issue, can you reference it here and I'll close this one?

**Comment by brendandburns** [comment]
(and concretely, I meant add retry in test-cmd.sh, not kubectl)

**Comment by deads2k** [comment]
> (and concretely, I meant add retry in test-cmd.sh, not kubectl)

To my knowledge, `kubectl patch` doesn't allow the user to specify a resource version.  In light of that, why not have `PatchResource` detect a conflict failure, build patch (call it changes) between old and current, check for conflicts between changes and user patch, if there are no conflicts, re-apply the user patch and issue the update?

That behavior is identical to what would happen if the user happened to had issued the `kubectl patch` an instant later, so it seems in keeping with user intent.

**Comment by smarterclayton** [comment]
Examples are:
https://github.com/openshift/origin/blob/master/test/cmd/deployments.sh#L41
and https://github.com/openshift/origin/blob/master/hack/util.sh#L178

The open issue was raised about precondition patch, not test-cmd retry.
It's probably a much bigger fix than what david suggested - this could
potentially address that.

**Comment by bgrant0607-nocc** [comment]
Patch preconditions mentioned here: https://github.com/kubernetes/kubernetes/issues/13576#issuecomment-137647830

**Comment by goltermann** [comment]
Some work happening at #14758

**Comment by brendandburns** [comment]
I'm closing this as I believe it is fixed by #14758
